File: measure.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"time"
)

func writeUsingFile(fileName string, numOfNums int) {
	file, _ := os.OpenFile(fileName, os.O_CREATE|os.O_WRONLY, 0600)
	defer file.Close()
	writer := bufio.NewWriter(file)
	for i := 1; i <= numOfNums; i++ {
		writer.WriteByte(byte(i & 0xff))
	}
}

func readUsingFile(fileName string, numOfNums int) {
	file, _ := os.Open(fileName)
	defer file.Close()
	reader := bufio.NewReader(file)
	for i := 1; i <= numOfNums; i++ {
		reader.ReadByte()
	}

	os.Remove(fileName)
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Provide filename")
		os.Exit(-1)
	}

	k := 256
	numOfNums := k * 1024 * 1024
	fileName := os.Args[1]
	reps := 6

	if os.Remove(fileName) != nil {
		// ignore
	}

	var read float32 = 0.0
	var write float32 = 0.0
	for i := 1; i <= reps; i++ {
		time1 := time.Now()
		writeUsingFile(fileName, numOfNums)
		time2 := time.Now()
		readUsingFile(fileName, numOfNums)
		time3 := time.Now()
		tmp1 := float32(numOfNums/1024/1024) * 1e9
		diff1 := time3.Sub(time2)
		readTime := tmp1 / float32(diff1)
		diff2 := time2.Sub(time1)
		writeTime := tmp1 / float32(diff2)
		fmt.Println(readTime, writeTime)
		if i != 1 {
			read += float32(readTime)
			write += float32(writeTime)
		}
	}

	tmp1 := reps - 1
	fmt.Println("Bytes:", numOfNums/1024/1024, "MiB, ReadRate:",
		read/float32(tmp1), "MB/s, WriteRate:", write/float32(tmp1), "MB/s")
}

File: measure.cpp
#include <chrono>
#include <fstream>
#include <iostream>
using namespace std;

int k = 256;
int numOfNums = k * 1024 * 1024;
int reps = 6;

void writeUsingFile(string fileName) {
  ofstream writer(fileName, ios_base::binary);
  for (int i = 1; i <= numOfNums; ++i)
    writer << (uint8_t) (i & 0xff);
}

void readUsingFile(string fileName) {
  ifstream reader(fileName, ios_base::binary);
  uint8_t c;
  for (int i = 1; i <= numOfNums; ++i) 
    reader >> c;
  remove(fileName.c_str());
}

int main(int argc, char* argv[]) {
  if (argc < 2) {
    cerr << "Provide filename" << endl;
    exit(-1);
  }
  string fileName(argv[1]);

  remove(fileName.c_str());

  float read = 0;
  float write = 0;
  for (int i = 1; i <= reps; i++) {
    auto time1 = chrono::system_clock::now();
    writeUsingFile(fileName);
    auto time2 = chrono::system_clock::now();
    readUsingFile(fileName);
    auto time3 = chrono::system_clock::now();
    auto tmp1 = numOfNums / 1024 / 1024 * 1e9;
    auto diff1 = chrono::duration_cast<chrono::nanoseconds>(time3 - time2).count();
    auto readTime = tmp1 / diff1;
    auto diff2 = chrono::duration_cast<chrono::nanoseconds>(time2 - time1).count();
    auto writeTime = tmp1 / diff2;
    cout << readTime << ' ' << writeTime << endl;
    if (i != 1) {
      read += readTime;
      write += writeTime;
    }
  } 
  int tmp1 = reps - 1;
  cout << "Bytes: " << numOfNums / 1024 / 1024 << " MiB, ReadRate: " << \
    read / tmp1 << "MB/s, WriteRate: " << write / tmp1 << " MB/s" << endl;
}
